src/module.ml

```diff
@@ -21,4 +22,21 @@ let cm_file t ~dir kind = Path.relative dir (t.obj_name ^ Cm_kind.ext kind)
 let cmt_file t ~dir (kind : Ml_kind.t) =
   match kind with
   | Impl -> Some (Path.relative dir (t.obj_name ^ ".cmt"))
-  | Intf -> Option.map t.mli_fname ~f:(fun _ -> Path.relative dir (t.obj_name ^ ".cmti"))
+  | Intf -> Option.map t.intf_fname ~f:(fun _ -> Path.relative dir (t.obj_name ^ ".cmti"))
+
+let create ?(obj_name="") ?intf_fname ~name ~impl_fname () =
+  match Filename.extension impl_fname with
+  | ".re" -> { obj_name ; name ; intf_fname ; impl_fname ; reason = true }
+  | ".ml" -> { obj_name ; name ; intf_fname ; impl_fname ; reason = false }
+  | _ -> invalid_arg ("unexpected extension in " ^ impl_fname)
+
+let ocaml_of_reason t =
+  if t.reason then
+    { t with
+      reason = false
+    ; impl_fname = Filename.chop_extension t.impl_fname ^ ".ml"
```

Comment:
Actually we should just keep the two extensions, i.e. the filename would be: `foo.re.ml`. The reason for this is that otherwise the generation of the rules is not a fix point: initially `guess_modules` picks up only `foo.re`, but if run again it would pickup `foo.re` and `foo.ml`. It doesn't matter right now, but it might for the jenga bridge. It might also matter for the implementation of #35.